codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/generators/protocol/RequestSerializerGenerator.kt

```diff
@@ -103,6 +116,59 @@ class RequestSerializerGenerator(
                 writable { }
             },
         )
+        writer.withBlockTemplate(
+            """
+            ##[derive(Debug)]
+            struct ${operationName}RequestSerializer;
+            impl #{RequestSerializer} for ${operationName}RequestSerializer {
+                ##[allow(unused_mut, clippy::let_and_return, clippy::needless_borrow, clippy::useless_conversion)]
+                fn serialize_input(&self, input: #{Input}) -> #{Future}<#{HttpRequest}> {
+            """,
+            "}}",
+            *fullCodegenScope,
+        ) {
+            val serializeInner = writable {
+                rustTemplate(
+                    """
+                    let mut request_builder = {
+                        #{create_http_request}
+                    };
+                    let body = #{generate_body};
+                    #{add_content_length}
+                    let mut request = request_builder.body(body).expect("valid request");
+                    """,
+                    *fullCodegenScope,
+                )
+            }
+            if (asyncAfterSerialization.isEmpty()) {
+                writer.rustTemplate(
+                    """
+                    fn serialize(input: #{ConcreteInput}) -> #{Result}<#{HttpRequest}, #{BoxError}> {
+                        #{serialize_inner}
+                        Ok(request)
+                    }
+                    let input = #{TypedBox}::<#{ConcreteInput}>::assume_from(input).expect("correct type").unwrap();
+                    #{Future}::ready(serialize(input))
+                    """,
+                    *fullCodegenScope,
+                    "serialize_inner" to serializeInner,
+                )
+            } else {
+                writer.rustTemplate(
+                    """
+                    async fn serialize(input: #{ConcreteInput}) -> #{Result}<#{HttpRequest}, #{BoxError}> {
+                        #{serialize_inner}
+                        #{async_after_serialization}
+                        Ok(request)
+                    }
+                    let input = #{TypedBox}::<#{ConcreteInput}>::assume_from(input).expect("correct type").unwrap();
```

Comment:
can we allow `#{async_after_serialization}` to be empty instead of having two different paths? Doesn't seem like having serialize be async given that the resulting generated state machine will only contain a ready variant is a problem

Reply:
Removed this entirely.